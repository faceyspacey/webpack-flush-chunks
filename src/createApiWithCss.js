// @flow
import React from 'react'
import fs from 'fs'

type StatelessComponent = () => React.Element<*>
type ObjectString = {
  toString: () => string
}
export type Api = {
  Js: StatelessComponent,
  Styles: StatelessComponent,
  Css: StatelessComponent,

  js: ObjectString,
  styles: ObjectString,
  css: ObjectString,

  scripts: Array<string>,
  stylesheets: Array<string>,

  publicPath: string,
  outputPath: ?string
}

const DEV = process.env.NODE_ENV === 'development'

/** CREATE API WITH CSS */

export default (
  files: Array<string>,
  filesOrderedForCss: Array<string>,
  publicPath: string,
  outputPath: ?string
): Api => {
  const regex = getJsFileRegex(files)
  const scripts = files.filter(file => isJs(regex, file))
  const stylesheets = filesOrderedForCss.filter(isCss)
  publicPath = publicPath.replace(/\/$/, '')

  const api = {
    // 1) Use as React components using ReactDOM.renderToStaticMarkup, eg:
    // <html><Styles /><Js /><html>
    Js: () => (
      <span>
        {scripts.map((file, key) => (
          <script
            type='text/javascript'
            src={`${publicPath}/${file}`}
            key={key}
          />
        ))}
      </span>
    ),
    Styles: () => (
      <span>
        {stylesheets.map((file, key) => (
          <link rel='stylesheet' href={`${publicPath}/${file}`} key={key} />
        ))}
      </span>
    ),

    // 2) Use as string, eg: `${styles} ${js}`
    js: {
      toString: () =>
        // lazy-loaded in case not used
        scripts
          .map(
            file =>
              `<script type='text/javascript' src='${publicPath}/${file}'></script>`
          )
          .join('\n')
    },
    styles: {
      toString: () =>
        // lazy-loaded in case not used
        stylesheets
          .map(file => `<link rel='stylesheet' href='${publicPath}/${file}' />`)
          .join('\n')
    },

    // 3) Embed the raw css without needing to load another file.
    // Use as a React component (<Css />) or a string (`${css}`):
    // NOTE: during development, HMR requires stylesheets.
    Css: () =>
      (DEV
        ? api.Styles()
        : <span>
          <style>{stylesAsString(stylesheets, outputPath)}</style>
        </span>),
    css: {
      toString: () =>
        // lazy-loaded in case not used
        (DEV
          ? api.styles.toString()
          : `<style>${stylesAsString(stylesheets, outputPath)}</style>`)
    },

    // 4) names of files without publicPath or outputPath prefixed:
    scripts,
    stylesheets,

    // 5) for completeness provide the paths even though they were inputs:
    publicPath,
    outputPath
  }

  return api
}

/** HELPERS */

const getJsFileRegex = (files: Array<string>): RegExp => {
  const isUsingExtractCssChunk = !!files.find(file => file.includes('no_css'))
  return isUsingExtractCssChunk ? /\.no_css\.js$/ : /\.js$/
}

const isJs = (regex: RegExp, file: string): boolean =>
  regex.test(file) && !/\.hot-update\.js$/.test(file)

const isCss = (file: string): boolean => /\.css$/.test(file)

const stylesAsString = (
  stylesheets: Array<string>,
  outputPath: ?string
): string => {
  if (!outputPath) {
    throw new Error(
      `No \`outputPath\` was provided as an option to \`flushChunks\`. 
      Please provide one so stylesheets can be read from the
      file system since you're embedding the css as a string.`
    )
  }

  const path = outputPath.replace(/\/$/, '')

  return stylesheets
    .map(file => {
      const filePath = `${path}/${file}`
      return fs.readFileSync(filePath, 'utf8')
    })
    .join('\n')
    .replace(/\/\*# sourceMappingURL=.+\*\//g, '') // hide prod sourcemap err
}

/** EXPORTS FOR TESTING */

export { getJsFileRegex, isJs, isCss, stylesAsString }
