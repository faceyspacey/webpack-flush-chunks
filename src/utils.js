export const isWebpack = () => typeof __webpack_require__ !== 'undefined'
export const requireReactWeakly = () => {
  if (!isWebpack()) {
    return module.require('react')
  }
  return __webpack_require__(require.resolveWeak('react'))
}

export default {
  isWebpack,
  requireReactWeakly
}
