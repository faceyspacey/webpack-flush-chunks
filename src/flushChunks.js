// @flow
import type { Api, CssChunksHash } from './createApiWithCss'
import type { Stats, Options } from './flushFiles'
import createApiWithCss from './createApiWithCss'
import { filesFromChunks, flush } from './flushFiles'

declare function __webpack_require__(id: string): any

const IS_WEBPACK = typeof __webpack_require__ !== 'undefined'
const IS_TEST = process.env.NODE_ENV === 'test'
const defaults = {
  before: ['bootstrap', 'vendor'],
  after: ['main']
}

/** PUBLIC API */

export default (stats: Stats, opts: Options): Api =>
  flushChunks(stats, IS_WEBPACK, opts)

const flushChunks = (stats: Stats, isWebpack: boolean, opts: Options = {}) => {
  const beforeEntries = opts.before || defaults.before
  const jsBefore = filesFromChunks(beforeEntries, stats.assetsByChunkName)

  const files = opts.chunkNames
    ? filesFromChunks(opts.chunkNames, stats.assetsByChunkName, true)
    : flush(opts.moduleIds || [], stats, opts.rootDir, isWebpack)

  const afterEntries = opts.after || defaults.after
  const jsAfter = filesFromChunks(afterEntries, stats.assetsByChunkName)

  return createApiWithCss(
    [...jsBefore, ...files, ...jsAfter],
    [
      ...jsBefore, // likely nothing in here, but if so: bootstrap.css, vendor.css, etc
      ...jsAfter.reverse(), // main.css, someElseYouPutBeforeMain.css, etc
      ...files // correct incrementing order already
    ],
    stats,
    opts.outputPath
  )
}

export { flushChunks }
export {
  flushFiles,
  flushFilesPure,
  flush,
  flushBabel,
  flushWebpack,
  createFilesByPath,
  createFilesByModuleId,
  isUnique,
  normalizePath,
  concatFilesAtKeys,
  filesFromChunks
} from './flushFiles'
