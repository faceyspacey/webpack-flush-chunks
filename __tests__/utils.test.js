import { requireReactWeakly } from '../src/utils'

describe('requireReactWeakly', () => {
  it('requires react module weakly', () => {
    const React = requireReactWeakly()
    /*
    This will cause warnings while running npm test since it access all properties of React to snapshot it and some properties as deprecated
    warnings:
    console.warn node_modules\react\lib\lowPriorityWarning.js:40
      Warning: Accessing PropTypes via the main React package is deprecated, and will be removed in  React v16.0. Use the latest available v15.* prop-types package from npm instead. For info on
 usage, compatibility, migration and more, see https://fb.me/prop-types-docs
    console.warn node_modules\react\lib\lowPriorityWarning.js:40
      Warning: Accessing createClass via the main React package is deprecated, and will be removed in React v16.0. Use a plain JavaScript class instead. If you're not yet ready to migrate, crea
te-react-class v15.* is available on npm as a temporary, drop-in replacement. For more info see https://fb.me/react-create-class
    */
    expect(React).toMatchSnapshot()
    /* ---- */
    /* that's the only function necessary */
    expect(React.createElement).toBeInstanceOf(Function)
  })
})
