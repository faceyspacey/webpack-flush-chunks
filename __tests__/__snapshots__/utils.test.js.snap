// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requireReactWeakly requires react module weakly 1`] = `
Object {
  "Children": Object {
    "count": [Function],
    "forEach": [Function],
    "map": [Function],
    "only": [Function],
    "toArray": [Function],
  },
  "Component": [Function],
  "DOM": Object {
    "a": [Function],
    "abbr": [Function],
    "address": [Function],
    "area": [Function],
    "article": [Function],
    "aside": [Function],
    "audio": [Function],
    "b": [Function],
    "base": [Function],
    "bdi": [Function],
    "bdo": [Function],
    "big": [Function],
    "blockquote": [Function],
    "body": [Function],
    "br": [Function],
    "button": [Function],
    "canvas": [Function],
    "caption": [Function],
    "circle": [Function],
    "cite": [Function],
    "clipPath": [Function],
    "code": [Function],
    "col": [Function],
    "colgroup": [Function],
    "data": [Function],
    "datalist": [Function],
    "dd": [Function],
    "defs": [Function],
    "del": [Function],
    "details": [Function],
    "dfn": [Function],
    "dialog": [Function],
    "div": [Function],
    "dl": [Function],
    "dt": [Function],
    "ellipse": [Function],
    "em": [Function],
    "embed": [Function],
    "fieldset": [Function],
    "figcaption": [Function],
    "figure": [Function],
    "footer": [Function],
    "form": [Function],
    "g": [Function],
    "h1": [Function],
    "h2": [Function],
    "h3": [Function],
    "h4": [Function],
    "h5": [Function],
    "h6": [Function],
    "head": [Function],
    "header": [Function],
    "hgroup": [Function],
    "hr": [Function],
    "html": [Function],
    "i": [Function],
    "iframe": [Function],
    "image": [Function],
    "img": [Function],
    "input": [Function],
    "ins": [Function],
    "kbd": [Function],
    "keygen": [Function],
    "label": [Function],
    "legend": [Function],
    "li": [Function],
    "line": [Function],
    "linearGradient": [Function],
    "link": [Function],
    "main": [Function],
    "map": [Function],
    "mark": [Function],
    "mask": [Function],
    "menu": [Function],
    "menuitem": [Function],
    "meta": [Function],
    "meter": [Function],
    "nav": [Function],
    "noscript": [Function],
    "object": [Function],
    "ol": [Function],
    "optgroup": [Function],
    "option": [Function],
    "output": [Function],
    "p": [Function],
    "param": [Function],
    "path": [Function],
    "pattern": [Function],
    "picture": [Function],
    "polygon": [Function],
    "polyline": [Function],
    "pre": [Function],
    "progress": [Function],
    "q": [Function],
    "radialGradient": [Function],
    "rect": [Function],
    "rp": [Function],
    "rt": [Function],
    "ruby": [Function],
    "s": [Function],
    "samp": [Function],
    "script": [Function],
    "section": [Function],
    "select": [Function],
    "small": [Function],
    "source": [Function],
    "span": [Function],
    "stop": [Function],
    "strong": [Function],
    "style": [Function],
    "sub": [Function],
    "summary": [Function],
    "sup": [Function],
    "svg": [Function],
    "table": [Function],
    "tbody": [Function],
    "td": [Function],
    "text": [Function],
    "textarea": [Function],
    "tfoot": [Function],
    "th": [Function],
    "thead": [Function],
    "time": [Function],
    "title": [Function],
    "tr": [Function],
    "track": [Function],
    "tspan": [Function],
    "u": [Function],
    "ul": [Function],
    "var": [Function],
    "video": [Function],
    "wbr": [Function],
  },
  "PropTypes": Object {
    "PropTypes": [Circular],
    "any": [Function],
    "array": [Function],
    "arrayOf": [Function],
    "bool": [Function],
    "checkPropTypes": [Function],
    "element": [Function],
    "func": [Function],
    "instanceOf": [Function],
    "node": [Function],
    "number": [Function],
    "object": [Function],
    "objectOf": [Function],
    "oneOf": [Function],
    "oneOfType": [Function],
    "shape": [Function],
    "string": [Function],
    "symbol": [Function],
  },
  "PureComponent": [Function],
  "__spread": [Function],
  "cloneElement": [Function],
  "createClass": [Function],
  "createElement": [Function],
  "createFactory": [Function],
  "createMixin": [Function],
  "isValidElement": [Function],
  "version": "15.6.1",
}
`;
